### PR TITLE
[EDA] Add Education History to  CON_CannotDelete_TDTM and ACCT_CannotDelete_TDTM classes

### LINKED ISSUE #1339: [EDA] Add Education History to  CON_CannotDelete_TDTM and ACCT_CannotDelete_TDTM classes
**Note: FOR INTERNAL USE ONLY. For support questions, please use the EDA Chatter group in [The Power of Us Hub](https://powerofus.force.com).
For feature requests, please use the Ideas tab in [The Power of Us Hub](https://powerofus.force.com).**

* **Do you want to request a *feature* or report a *bug*?**
feature
* **What is the current behavior?**
CON_CannotDelete_TDTM and ACCT_CannotDelete_TDTM classes prevent Account and Contact from being deleted when they have child record association and Prevent Deletion flags are enabled. 

* **What is the expected behavior?**
Updated CON_CannotDelete_TDTM and ACCT_CannotDelete_TDTM to include Education History to the existing Prevent Delete logic. With this change an Account or a Contact with Education History as one of the related records can not be deleted when Prevent Deletion flags are enabled.

* **What is the motivation / use case for changing the behavior?**
Scale the functionality to Education History

* **Acceptance Criteria**

  - **Given**: When I am on a Account And Contact Page
  - **When**: Prevent Deletion flags are enabled
  - **Then**: Deleting the Account and Contact records with EducationHistory related records should prevent deletion by throwing an error.

### PR DESCRIPTION
**Pull request by skosgi1**
# Critical Changes

# Changes
We've updated CON_CannotDelete_TDTM and ACCT_CannotDelete_TDTM to include Education History to the existing Prevent Delete logic. With this change an Account or a Contact with Education History as one of the related records can not be deleted when Prevent Deletion flags are enabled.

# Issues Closed
Closes #1339 

# New Metadata

# Deleted Metadata

# Testing Notes

### CODE REVIEW

File: src/classes/ACCT_CannotDelete_TDTM.cls [hunk 1]
```diff
@@ -34,67 +34,78 @@
...
                                 (SELECT ID FROM Account.Program_Plans__r LIMIT 1),
                                 (SELECT ID FROM Account.Terms__r LIMIT 1),
                                 (SELECT ID FROM Account.Time_Blocks__r LIMIT 1)
-                                FROM Account WHERE ID in :oldlist]) {
-                
-                if (hasChildRecords(a)) {
-                    Account accountInContext = oldMap.get(a.ID);
+                                FROM Account WHERE ID IN :oldlist]) {
+
+                if (hasChildRecords(acc)) {
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TDTM.cls:82]: Let's keep using this.methodname
```suggestion
                if (this.hasChildRecords(acc)) {
```

File: src/classes/ACCT_CannotDelete_TDTM.cls [hunk 2]
```diff
@@ -34,67 +34,78 @@
...
+                acc.Courses__r.size() > 0 ||
+                acc.Course_Enrollments__r.size() > 0 ||
+                acc.Education_History__r.size() > 0 ||
+                acc.Facilities__r.size() > 0 ||
+                acc.Household_Members__r.size() > 0 ||
+                acc.Organization_Members__r.size() > 0 ||
+                acc.Program_Enrollments__r.size() > 0 ||
+                acc.Program_Plans__r.size() > 0 ||
+                acc.Terms__r.size() > 0 ||
+                acc.Time_Blocks__r.size() > 0);
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TDTM.cls:109]: We should be using isEmpty(). Splitting this out across each line made this much easier to read.
```suggestion
        return (acc.Addresses__r.isEmpty() == false ||
                acc.Affl_Contacts__r.isEmpty() == false ||
                acc.Attributes__r.isEmpty() == false ||
                acc.Courses__r.isEmpty() == false ||
                acc.Course_Enrollments__r.isEmpty() == false ||
                acc.Education_History__r.isEmpty() == false ||
                acc.Facilities__r.isEmpty() == false ||
                acc.Household_Members__r.isEmpty() == false ||
                acc.Organization_Members__r.isEmpty() == false ||
                acc.Program_Enrollments__r.isEmpty() == false ||
                acc.Program_Plans__r.isEmpty() == false ||
                acc.Terms__r.isEmpty() == false ||
                acc.Time_Blocks__r.isEmpty() == false);
```

File: src/classes/ACCT_CannotDelete_TEST.cls [hunk 3]
```diff
@@ -295,7 +296,39 @@ public with sharing class ACCT_CannotDelete_TEST {
...
+    public static void accCannotDeleteWithEducationHistoryBULK() {
+        UTIL_CustomSettingsFacade.getSettingsForTests(new Hierarchy_Settings__c
+                                                            (Account_Processor__c = bizAccRecordTypeId,
+                                                            Prevent_Account_Deletion__c = true));
+
+        List<Account> newAccounts = UTIL_UnitTestData_TEST.getMultipleTestAccounts(2, UTIL_Describe.getAcademicAccRecTypeID()); 
+        insert newAccounts;
+        
+        Education_History__c educationHistory1 = UTIL_UnitTestData_TEST.getEduHistory(newAccounts[0].Id, null);
+        Education_History__c educationHistory2 = UTIL_UnitTestData_TEST.getEduHistory(newAccounts[1].Id, null);
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TEST.cls:314]: Please add more than one education history per account. Even though that is not our standard use case, we should use a multiplicity of records at the parent and child relationship we are testing against.

File: src/classes/ACCT_CannotDelete_TEST.cls [hunk 4]
```diff
@@ -701,7 +734,38 @@ public with sharing class ACCT_CannotDelete_TEST {
...
+    public static void accCanDeleteWithEducationHistoryBULK() {
+
+        UTIL_CustomSettingsFacade.getSettingsForTests(new Hierarchy_Settings__c
+                                                            (Account_Processor__c = bizAccRecordTypeId,
+                                                            Prevent_Account_Deletion__c = false));
+
+        List<Account> newAccounts = UTIL_UnitTestData_TEST.getMultipleTestAccounts(2, UTIL_Describe.getAcademicAccRecTypeID()); 
+        insert newAccounts;
+
+        Education_History__c educationHistory1 = UTIL_UnitTestData_TEST.getEduHistory(newAccounts[0].Id, null);
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TEST.cls:752]: Likewise here.

File: src/classes/ACCT_CannotDelete_TEST.cls [hunk 5]
```diff
@@ -961,7 +1025,10 @@ public with sharing class ACCT_CannotDelete_TEST {
         insert term; 
 
         Time_Block__c timeBlock = new Time_Block__c (Name = 'TimeB', Educational_Institution__c = lastNameByContact.get('Willy').AccountId);
-        insert timeBlock; 
+        insert timeBlock;
+
+        Education_History__c educationHistory = UTIL_UnitTestData_TEST.getEduHistory(lastNameByContact.get('Willy').AccountId, null);
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TEST.cls:1030]: Note that for the large-scale integration test, one record is fine, unless we want to build a engine to provide multiple records at each phase.

File: src/classes/ACCT_CannotDelete_TEST.cls [hunk 6]
```diff
@@ -1022,7 +1089,10 @@ public with sharing class ACCT_CannotDelete_TEST {
         insert course; 
 
         Facility__c facility = new Facility__c (Name = 'MeetingRoom1', Account__c = lastNameByContact.get('Willy').AccountId);
-        insert facility; 
+        insert facility;
+
+        Education_History__c educationHistory = UTIL_UnitTestData_TEST.getEduHistory(lastNameByContact.get('Willy').AccountId, null);
```
Comment (MatthewBlanski) [inline src/classes/ACCT_CannotDelete_TEST.cls:1094]: Likewise here for the integration scale.

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
One or two very minor changes around styling and around bulk test data, then we're good to go.

**Review by MatthewBlanski — APPROVED** [review]

**Comment by spinkelman** [comment]
QA Passes - Merging
Will Close WI

**Pull request merged**